File: worker/app/highlight_runner.py
"""Highlight render runner.

Pops a highlight_id (via worker.py's highlight loop), cuts a clip around each
selected event, then stitches them into one MP4 with ffmpeg concat. Mirrors the
VisionJob lifecycle in vision_runner: status pending -> running -> done/failed,
with per-clip progress.
"""
from datetime import datetime, timezone
from pathlib import Path

from sqlalchemy import select

from app.config import get_settings
from app.db import async_session_maker
from app.models import Event, Highlight, HighlightStatus, Video

settings = get_settings()


def _parse_event_ids(csv: str) -> list[int]:
    out: list[int] = []
    for part in (csv or "").split(","):
        part = part.strip()
        if part:
            try:
                out.append(int(part))
            except ValueError:
                pass
    return out


async def run_highlight_job(highlight_id: int) -> None:
    """Renders a highlight reel from its selected events. Runs in the worker thread."""
    from vision.clips import concat_clips, extract_clip

    async with async_session_maker() as session:
        highlight = await session.get(Highlight, highlight_id)
        if highlight is None:
            return

        highlight.status = HighlightStatus.running
        highlight.progress = 0.0
        await session.commit()

        try:
            event_ids = _parse_event_ids(highlight.event_ids)
            if not event_ids:
                raise ValueError("No events selected")

            # Fetch the selected events, ordered by timestamp so the reel plays in
            # match order. Preserve only events from this analysis.
            rows = (
                await session.execute(
                    select(Event)
                    .where(Event.analysis_id == highlight.analysis_id)
                    .where(Event.id.in_(event_ids))
                    .order_by(Event.timestamp_seconds)
                )
            ).scalars().all()
            if not rows:
                raise ValueError("Selected events not found")

            # Cache video file paths by id (events may span multiple source videos).
            video_paths: dict[int, str] = {}

            clip_paths: list[Path] = []
            total = len(rows)
            for i, ev in enumerate(rows):
                if ev.video_id is None:
                    continue
                if ev.video_id not in video_paths:
                    video = await session.get(Video, ev.video_id)
                    if video is None:
                        continue
                    video_paths[ev.video_id] = str(settings.uploads_dir / video.filename)

                start = max(0.0, ev.timestamp_seconds - settings.clip_pre_seconds)
                end = ev.timestamp_seconds + settings.clip_post_seconds
                clip_name = f"hl{highlight.id}_ev{ev.id}.mp4"
                clip_path = settings.clips_dir / clip_name
                if extract_clip(video_paths[ev.video_id], clip_path, start, end):
                    clip_paths.append(clip_path)

                # progress is the clipping phase (0..0.8); concat is the final 0.8..1.0
                highlight.progress = round(0.8 * (i + 1) / total, 3)
                await session.commit()

            if not clip_paths:
                raise ValueError("Could not extract any clips")

            out_name = f"highlight{highlight.id}.mp4"
            out_path = settings.clips_dir / out_name
            if not concat_clips(clip_paths, out_path):
                raise RuntimeError("ffmpeg concat failed")

            highlight.filename = out_name
            highlight.status = HighlightStatus.done
            highlight.progress = 1.0
            highlight.finished_at = datetime.now(timezone.utc)
            await session.commit()

        except Exception as exc:  # noqa: BLE001
            await session.rollback()
            highlight = await session.get(Highlight, highlight_id)
            if highlight:
                highlight.status = HighlightStatus.failed
                highlight.error = f"{type(exc).__name__}: {exc}"
                highlight.finished_at = datetime.now(timezone.utc)
                await session.commit()

File: worker-dotnet/PitchWise.Worker/HighlightRunner.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using PitchWise.Api.Config;
using PitchWise.Api.Data;
using PitchWise.Api.Models;
using PitchWise.Vision;

namespace PitchWise.Worker;

/// <summary>
/// Renders a highlight reel: cut a clip around each selected event, concat into one MP4,
/// then HLS-segment. Port of worker/app/highlight_runner.py run_highlight_job.
/// </summary>
public sealed class HighlightRunner
{
    private readonly IServiceScopeFactory _scopeFactory;
    private readonly AppSettings _appSettings;
    private readonly WorkerSettings _worker;

    public HighlightRunner(
        IServiceScopeFactory scopeFactory, AppSettings appSettings, WorkerSettings worker)
    {
        _scopeFactory = scopeFactory;
        _appSettings = appSettings;
        _worker = worker;
    }

    private static List<int> ParseEventIds(string csv) =>
        (csv ?? "").Split(',')
            .Select(s => s.Trim())
            .Where(s => s.Length > 0 && int.TryParse(s, out _))
            .Select(int.Parse)
            .ToList();

    public async Task RunAsync(int highlightId, CancellationToken ct)
    {
        using IServiceScope scope = _scopeFactory.CreateScope();
        AppDbContext db = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        Highlight? highlight = await db.Highlights.FirstOrDefaultAsync(h => h.Id == highlightId, ct);
        if (highlight is null) return;

        highlight.Status = HighlightStatus.Running;
        highlight.Progress = 0.0;
        await db.SaveChangesAsync(ct);

        try
        {
            List<int> eventIds = ParseEventIds(highlight.EventIds);
            if (eventIds.Count == 0) throw new InvalidOperationException("No events selected");

            // Selected events for this analysis, in match order.
            var rows = await db.Events
                .Where(e => e.AnalysisId == highlight.AnalysisId && eventIds.Contains(e.Id))
                .OrderBy(e => e.TimestampSeconds)
                .ToListAsync(ct);
            if (rows.Count == 0) throw new InvalidOperationException("Selected events not found");

            var videoPaths = new Dictionary<int, string>();
            var clipPaths = new List<string>();
            int total = rows.Count;
            for (int i = 0; i < rows.Count; i++)
            {
                Event ev = rows[i];
                if (ev.VideoId is not int vid) continue;
                if (!videoPaths.TryGetValue(vid, out string? vpath))
                {
                    Video? video = await db.Videos.FirstOrDefaultAsync(v => v.Id == vid, ct);
                    if (video is null) continue;
                    vpath = Path.Combine(_appSettings.UploadsDir, video.Filename);
                    videoPaths[vid] = vpath;
                }

                double start = Math.Max(0.0, ev.TimestampSeconds - _worker.ClipPreSeconds);
                double end = ev.TimestampSeconds + _worker.ClipPostSeconds;
                string clipName = $"hl{highlight.Id}_ev{ev.Id}.mp4";
                string clipPath = Path.Combine(_appSettings.ClipsDir, clipName);
                if (FfmpegTools.ExtractClip(vpath, clipPath, start, end))
                    clipPaths.Add(clipPath);

                highlight.Progress = Math.Round(0.8 * (i + 1) / total, 3);
                await db.SaveChangesAsync(ct);
            }

            if (clipPaths.Count == 0) throw new InvalidOperationException("Could not extract any clips");

            string outName = $"highlight{highlight.Id}.mp4";
            string outPath = Path.Combine(_appSettings.ClipsDir, outName);
            if (!FfmpegTools.ConcatClips(clipPaths, outPath))
                throw new InvalidOperationException("ffmpeg concat failed");

            highlight.Filename = outName;
            highlight.Status = HighlightStatus.Done;
            highlight.Progress = 1.0;
            highlight.FinishedAt = DateTime.UtcNow;

            // Best-effort HLS; MP4 fallback still works if it fails.
            string hlsDir = Path.Combine(_appSettings.HlsDir, highlight.Id.ToString());
            if (FfmpegTools.ToHls(outPath, hlsDir))
                highlight.HlsReady = true;

            await db.SaveChangesAsync(ct);
        }
        catch (Exception exc)
        {
            Highlight? h = await db.Highlights.FirstOrDefaultAsync(x => x.Id == highlightId, ct);
            if (h is not null)
            {
                h.Status = HighlightStatus.Failed;
                h.Error = $"{exc.GetType().Name}: {exc.Message}";
                h.FinishedAt = DateTime.UtcNow;
                await db.SaveChangesAsync(ct);
            }
        }
    }
}
